staticaddr/deposit/fsm.go

```diff
@@ -0,0 +1,303 @@
+
+	Failed = fsm.StateType("DepositFailed")
+)
+
+// Events.
+var (
+	OnStart           = fsm.EventType("OnStart")
+	OnExpiry          = fsm.EventType("OnExpiry")
+	OnExpiryPublished = fsm.EventType("OnExpiryPublished")
+	OnExpirySwept     = fsm.EventType("OnExpirySwept")
+	OnRecover         = fsm.EventType("OnRecover")
+)
+
+// FSM is the state machine that handles the instant out.
+type FSM struct {
+	*fsm.StateMachine
+
+	cfg *ManagerConfig
+
+	deposit *Deposit
+
+	params *address.Parameters
+
+	address *script.StaticAddress
+
+	ctx context.Context
+
+	blockNtfnChan chan uint32
+
+	finalizedDepositChan chan wire.OutPoint
+}
+
+// NewFSM creates a new state machine that can action on all static address
+// feature requests.
+func NewFSM(ctx context.Context, deposit *Deposit, cfg *ManagerConfig,
+	finalizedDepositChan chan wire.OutPoint,
+	recoverStateMachine bool) (*FSM, error) {
+
+	params, err := cfg.AddressManager.GetStaticAddressParameters(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get static address "+
+			"parameters: %v", err)
+	}
+
+	address, err := cfg.AddressManager.GetStaticAddress(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get static address: %v", err)
+	}
+
+	depoFsm := &FSM{
+		cfg:                  cfg,
+		deposit:              deposit,
+		params:               params,
+		address:              address,
+		ctx:                  ctx,
+		blockNtfnChan:        make(chan uint32),
+		finalizedDepositChan: finalizedDepositChan,
+	}
+
+	depositStates := depoFsm.DepositStatesV0()
+	switch params.ProtocolVersion {
+	case staticaddr.ProtocolVersion_V0:
+
+	default:
+		return nil, ErrProtocolVersionNotSupported
+	}
+
+	if recoverStateMachine {
+		depoFsm.StateMachine = fsm.NewStateMachineWithState(
+			depositStates, deposit.State,
+			DefaultObserverSize,
+		)
+	} else {
+		depoFsm.StateMachine = fsm.NewStateMachine(
+			depositStates, DefaultObserverSize,
+		)
+	}
+
+	depoFsm.ActionEntryFunc = depoFsm.updateDeposit
+
+	go func() {
+		for {
+			select {
+			case currentHeight := <-depoFsm.blockNtfnChan:
+				err := depoFsm.handleBlockNotification(
+					currentHeight,
+				)
+				if err != nil {
+					log.Errorf("error handling block "+
+						"notification: %v", err)
+				}
+
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+
+	return depoFsm, nil
+}
+
+// handleBlockNotification inspects the current block height and sends the
+// OnExpiry event to publish the expiry sweep transaction if the deposit timed
+// out, or it republishes the expiry sweep transaction if it was not yet swept.
+func (f *FSM) handleBlockNotification(currentHeight uint32) error {
+	params, err := f.cfg.AddressManager.GetStaticAddressParameters(f.ctx)
+	if err != nil {
+		return err
+	}
+
+	// If the deposit is expired but not yet sufficiently confirmed, we
+	// republish the expiry sweep transaction.
+	if f.deposit.isExpired(currentHeight, params.Expiry) {
+		if f.deposit.State == WaitForExpirySweep {
```

Comment:
>  we could just send the onexpiry event and in the fsm call the action itself again when the event is sent.

Wouldn't the new merged state action also wait for the confirmation of the sweep and hence block processing of new events until the state action returns a.k.a the expiry got confirmed? So we couldn't call `buildAndPublishTimeoutSweep` again since we wouldn't receive the event till the sweep is confirmed.

What do you think about extracting the sweep tx once we are in the final state `Expired` and then update the deposit with it, rather than needing to keep track of replacement txns while the sweep is unconfirmed?

Reply:
As a side note this is also not safe as we're running on the block notification goroutine.

Reply:
Thanks yes, I've added a mutex to the deposit and locked all operations on the state.